src/lib/device-online-state.ts

```diff
@@ -0,0 +1,340 @@
+import * as Bluebird from 'bluebird';
+import { DEFAULT_SUPERVISOR_POLL_INTERVAL } from './env-vars';
+import { noop } from 'lodash';
+import { resinApi, root } from '../platform';
+import { captureException } from '../platform/errors';
+import {
+	createPromisifedRedisClient,
+	PromisifedRedisClient,
+} from './redis-promise';
+import * as RedisSMQ from 'rsmq';
+import {
+	REDIS_HOST,
+	REDIS_PORT,
+	API_HEARTBEAT_STATE_ENABLED,
+	API_HEARTBEAT_STATE_TIMEOUT_SECONDS,
+} from './config';
+
+const getPollIntervalForDevice = resinApi.prepare<{ uuid: string }>({
+	resource: 'device_config_variable',
+	options: {
+		$select: 'value',
+		$top: 1,
+		$expand: {
+			device: {
+				$filter: { '@': 'uuid' },
+			},
+		},
+		$filter: {
+			name: {
+				$in: [
+					'BALENA_SUPERVISOR_POLL_INTERVAL',
+					'RESIN_SUPERVISOR_POLL_INTERVAL',
+				],
+			},
+		},
+		$orderby: {
+			// we want the last value that would have been passed
+			// to the supervisor, as that is the one it would have used.
+			name: 'desc',
+		},
+	},
+});
+
+const getPollIntervalForParentApplication = resinApi.prepare<{ uuid: string }>({
+	resource: 'application_config_variable',
+	options: {
+		$select: 'value',
+		$top: 1,
+		$filter: {
+			application: {
+				$any: {
+					$alias: 'a',
+					$expr: {
+						a: {
+							owns__device: {
+								$any: {
+									$alias: 'd',
+									$expr: {
+										d: {
+											uuid: { '@': 'uuid' },
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			name: {
+				$in: [
+					'BALENA_SUPERVISOR_POLL_INTERVAL',
+					'RESIN_SUPERVISOR_POLL_INTERVAL',
+				],
+			},
+		},
+		$orderby: {
+			// we want the last value that would have been passed
+			// to the supervisor, as that is the one it would have used.
+			name: 'desc',
+		},
+	},
+});
+
+export const getPollInterval = (uuid: string): Bluebird<number> => {
+	return getPollIntervalForDevice({ uuid }, undefined, { req: root })
+		.then(([pollInterval]: Array<{ value: number }>) => {
```

Comment:
This would be nicer as
```suggestion
		.then((pollIntervals: Array<{ value: number }>) => {
			if (pollIntervals.length > 0) {
				return pollIntervals;
			}
```
to avoid the need to destructure and then restructure the array